### PR TITLE
Support Qwix quantization on NNX

### PR DESCRIPTION
**Pull request by hsuan-lun-chiang**
# Description

This PR enables full support for Qwix quantization algorithms on NNX models and restores support for Linen models. Because the underlying Transformer architecture has been migrated to an NNX module, the existing Qwix quantization integration for Linen models was broken. This PR fixes the Linen integration and extends full feature parity to the NNX backend.

# Tests

Gpt3-6b
[Qwix Quantization Test - Linen](https://paste.googleplex.com/5058255918333952)
[Qwix Quantization Test - NNX](https://paste.googleplex.com/6613645239910400)

Gemma4-26b
[Qwix Quantization Test - Linen](https://paste.googleplex.com/5045455372288000)
[Qwix Quantization Test - NNX](https://paste.googleplex.com/5499316176289792)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/nnx_wrappers.py [hunk 1]
```diff
@@ -499,7 +499,16 @@ def maybe_unbox(x):
...
         paths_str += f"\n  - {'/'.join(map(str, path))}"
 
-      warnings.warn(f"Found unknown module paths in incoming state:{paths_str}")
+        # Dynamically reconstruct the unknown variables
+        curr = module
+        for p in path[:-1]:
+          if not hasattr(curr, p):
+            setattr(curr, p, nnx.Module())
+          curr = getattr(curr, p)
+
```
[the remark this answers is not in the document]
Comment (hsuan-lun-chiang) [reply src/maxtext/layers/nnx_wrappers.py:537]: Updated the dynamic reconstruction logic in ToLinen. The loop now explicitly checks isinstance(curr, dict), isinstance(curr, list), and isinstance(curr, tuple) to construct the missing paths safely without throwing  hasattr  errors on integer indices.

File: src/maxtext/trainers/pre_train/train.py [hunk 2]
```diff
@@ -371,15 +374,18 @@ def train_step(model, config, state_mesh_shardings, params_shardings, state, dat
...
+        _, _, new_overwrite_vars, new_rest = nnx.split(local_model, nnx.Param, OverwriteWithGradient, ...)
+        return loss, (aux, new_overwrite_vars, new_rest)
 
-      grad_func = jax.value_and_grad(diff_wrapper, argnums=0, has_aux=True)
-      (loss, (aux, new_rest)), raw_grads = grad_func(curr_params, rest, config, data)
+      grad_func = jax.value_and_grad(diff_wrapper, argnums=(0, 1), has_aux=True)
+      (loss, (aux, _, new_rest)), (raw_grads, overwrite_grads) = grad_func(
+          curr_params, overwrite_vars, rest, config, data
+      )
       nnx.update(state.model, new_rest)
```
[the remark this answers is not in the document]
Comment (hsuan-lun-chiang) [reply src/maxtext/trainers/pre_train/train.py]: By updating with overwrite_grads, we hook into JAX's existing gradient accumulation and distributed all-reduce infrastructure. This ensures that statistics like AMAX are automatically and correctly aggregated across all data-parallel workers, rather than having to manually handle the all-reduce logic for new_overwrite_vars. I've added a comment in train_step right above the nnx.update call to clarify this behaviour.

File: src/maxtext/layers/quantizations.py [hunk 3]
```diff
@@ -707,6 +711,28 @@ def configure_kv_quant(config):
...
+  """Applies a Linen module within an NNX context."""
+  try:
+    parent = flax_util.get_current_module()
+    is_nnx = isinstance(parent, nnx.Module)
+  except Exception:  # pylint: disable=broad-exception-caught
+    is_nnx = False
+
+  if is_nnx:
+    attr_name = f"_qwix_fp8_gpu_{op_id}"
+    if not hasattr(parent, attr_name):
```
[the remark this answers is not in the document]
Comment (hsuan-lun-chiang) [reply src/maxtext/layers/quantizations.py:726]: Updated the logic to check if  parent.rngs  exists (which is the standard NNX convention). If it does, we now call  parent.rngs.fork()  to cleanly branch the RNG stream for the quantization layer. If no parent rngs object is found, it safely falls back to nnx.Rngs(0) .

File: src/maxtext/layers/quantizations.py [hunk 4]
```diff
@@ -812,7 +838,21 @@ def maybe_quantize_model(model, config):
   if config.use_qwix_quantization and not config.use_batch_split_schedule:
     quantization_provider = get_qt_provider(config)
     if quantization_provider:
-      model = qwix.quantize_model(model, quantization_provider)
+      if config.pure_nnx:
+        input_shape = (config.micro_batch_size_to_train_on, config.max_target_length)
+        dummy_tokens = jnp.ones(input_shape, dtype=jnp.int32)
+        dummy_positions = jnp.ones(input_shape, dtype=jnp.int32)
+        dummy_segment_ids = jnp.ones(input_shape, dtype=jnp.int32)
+        model = qwix.quantize_model(
```
[the remark this answers is not in the document]
Comment (hsuan-lun-chiang) [reply src/maxtext/layers/quantizations.py:852]: They are purely dummy values for tracing shapes, so the jnp.ones array of the appropriate token shape is sufficient.

File: src/maxtext/utils/model_creation_utils.py [hunk 5]
```diff
@@ -534,7 +534,10 @@ def get_nnx_create_model_fn(config, mesh=None, devices=None, model_mode=MODEL_MO
 
   def _create_model():
     rngs = maxtext_utils_nnx.create_nnx_rngs(config, model_mode=model_mode, rng_key=rng_key)
-    return from_config(config, devices, mesh, rngs=rngs, model_mode=model_mode)
+    model = from_config(config, devices, mesh, rngs=rngs, model_mode=model_mode)
+    if config.pure_nnx:
+      _ = nnx.pop(model, nnx.Intermediate)
```
Comment (bvandermoon) [inline src/maxtext/utils/model_creation_utils.py]: Why are we popping the intermediates here?

File: src/maxtext/layers/quantizations.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bvandermoon) [reply src/maxtext/layers/quantizations.py:852]: If they are just for tracing shapes, why not use `eval_shape`?

[the remark this answers is not in the document]
Comment (hsuan-lun-chiang) [reply src/maxtext/layers/quantizations.py:852]: Because qwix.quantize_model actually executes a forward pass to trace the model and initialize the quantization states, we must provide actual numerical arrays (dummy variables). If we use eval_shape or  jax.ShapeDtypeStruct, it fails because it only provides shape metadata and cannot be used in actual tensor operations during the forward pass.

File: src/maxtext/utils/model_creation_utils.py (hunk 5 above)
Comment (hsuan-lun-chiang) [reply src/maxtext/utils/model_creation_utils.py]: When `create_nnx_abstract_model` traces the model shapes, the forward pass naturally generates these intermediates. We must pop them here so the abstract state perfectly matches the clean, persistent parameter tree that is saved to and loaded from checkpoints. If we didn't pop them here, the system would mistakenly expect them to exist during checkpoint restoration, leading to PyTree mismatch errors.

**Review by bvandermoon — COMMENTED** [review]
What is the `NaN` in the last line of https://paste.googleplex.com/5058255918333952?

**Comment by RexBearIU** [comment]
> What is the `NaN` in the last line of https://paste.googleplex.com/5058255918333952?

 The warning `NaN or Inf found in input tensor` in `x2num.py` is a benign TensorBoard logging warning caused by the perplexity metric overflowing, not by any actual NaN/Inf in model weights or gradients.
  
 Since perplexity is calculated as `exp(loss)`, and at the start of training the loss is extremely high (above 88.7), `exp(loss)` overflows to positive infinity (`inf`) in standard float32 precision. When TensorBoard's       
  summary writer attempts to log this infinity value, its internal validation check in `x2num.py` raises the warning.
  
 This is completely normal and occurs identically in unquantized pre-training. Here is a log snippet from a standard, unquantized baseline pre-training run showing the exact same loss progression and warning:
           
```text
    completed step: 0, loss: 445.785, lm_loss: 445.785, perplexity: inf 
    W0610 08:12:16.985936 x2num.py:13] NaN or Inf found in input tensor.
  
    completed step: 1, loss: 445.785, lm_loss: 445.785, perplexity: inf 
    W0610 08:12:18.443920 x2num.py:13] NaN or Inf found in input tensor.
  
    completed step: 2, loss: 217.815, lm_loss: 217.815, perplexity: inf 
    W0610 08:12:19.904166 x2num.py:13] NaN or Inf found in input tensor.
```

File: src/maxtext/layers/nnx_decoders.py [hunk 6]
```diff
@@ -536,22 +536,22 @@ def _update_leaf(leaf):
...
 
-    graphdef, state = nnx.split(layer)
-
-    def pure_layer_fn(state_in, y_in):
-      merged_layer = nnx.merge(graphdef, state_in)
-      out = merged_layer(y_in, **kwargs)
-      return out, nnx.state(merged_layer)
-
     # Linen FP8 ops keep amax_history in mutable Linen scope; jax.checkpoint
     # re-traces and hits UnexpectedTracerError. Skip remat for FP8.
```
Comment (bvandermoon) [inline src/maxtext/layers/nnx_decoders.py]: Why are we skipping remat for `FP8`? This is different from Linen, right?

Comment (RexBearIU) [reply src/maxtext/layers/nnx_decoders.py]: Under Flax Linen, `nn.remat` is a custom wrapper provided by Flax that has specialized handling for Flax's internal variable collections (such as the `amax_history` mutable collection used by FP8 operations) during rematerialization.

Under the NNX decoder when bridging NNX to Linen, the `_apply_layer_with_remat` helper splits the NNX module state and wraps the pure functionalized pass in standard `jax.checkpoint`. Since standard `jax.checkpoint` does not possess Flax Linen's specialized state-tracking mechanisms, the re-tracing process intercepts Flax's mutable tracers (which are captured inside Flax's `Scope` object), resulting in `UnexpectedTracerError`.

Therefore, we must skip the standard `jax.checkpoint` rematerialization when using Linen-based FP8 quantization configurations (`fp8_nanoo`, `fp8_gpu`) inside the NNX framework to avoid this tracing issue. This is indeed a difference between pure Linen and NNX decoder checkpointing.

Comment (bvandermoon) [reply src/maxtext/layers/nnx_decoders.py]: @cgarciae can you please take a look at this? Is there a way to make this work from the NNX persepctive? I don't think we will be able to complete the migration without FP8 remat working

File: src/maxtext/layers/nnx_wrappers.py [hunk 7]
```diff
@@ -161,6 +161,33 @@ def lazy_init(fn: Module | tp.Callable[..., tp.Any], *args, **kwargs):
   return fn
 
 
+def isolate_linen_stacks(fn: tp.Callable[..., tp.Any], *args, **kwargs):
```
Comment (cgarciae) [inline src/maxtext/layers/nnx_wrappers.py]: what is this doing?

Comment (RexBearIU) [reply src/maxtext/layers/nnx_wrappers.py]: **What `isolate_linen_stacks` does and why it's needed:**
  
During JAX compilation (`.init` and `.apply`), Flax Linen uses a thread-local stack (`linen.module._context.module_stack`) to resolve the hierarchy and scopes of active Linen modules.
  
However, on our **Flax NNX** path, we dynamically execute wrapped Linen operators (via `ToNNX`) inside an outer NNX model context. If the outer  context or JAX transforms have active module contexts, the thread-local Linen stack can retain stale/unexpected elements, leading to compiled tracer leaks and JAX `UnexpectedTracerError`.
  
`isolate_linen_stacks` solves this by temporarily shadowing and clearing the thread-local Linen and Bridge module stacks during the inner `ToNNX`  execution, and restoring them immediately after.

Comment (cgarciae) [reply src/maxtext/layers/nnx_wrappers.py]: `init_with_output` already does this isolation though...

Comment (RexBearIU) [reply src/maxtext/layers/nnx_wrappers.py]: You're totally right, good catch. I overlooked that `init_with_output` already handles that isolation under the hood. I've removed the redundant wrapper.

File: src/maxtext/layers/quantizations.py [hunk 8]
```diff
@@ -710,6 +713,32 @@ def configure_kv_quant(config):
   return None if not config.quantize_kvcache else KVQuant(config)
 
 
+def _apply_linen_module_in_nnx(linen_module_cls, op_id, *args, **kwargs):
+  """Applies a Linen module within an NNX context."""
+  try:
+    parent = flax_util.get_current_module()
+    is_nnx = isinstance(parent, nnx.Module)
+  except Exception:  # pylint: disable=broad-exception-caught
+    is_nnx = False
```
Comment (bvandermoon) [inline src/maxtext/layers/quantizations.py]: What is the exception being caught here? Is it `ValueError` from Qwix? Mind just catching the specific error if so?

**Review by bvandermoon — APPROVED** [review]

File: src/maxtext/layers/quantizations.py (hunk 8 above)
Comment (RexBearIU) [reply src/maxtext/layers/quantizations.py]: Thanks to point this out! When called outside of a module context, `flax_util.get_current_module()` raises a `ValueError` (`"Current module is not known."`). I have updated the block to catch `ValueError` explicitly.

**Review by NuojCheng — APPROVED** [review]

**Review by khatwanimohit — APPROVED** [review]
